### DISCUSSION TITLE
'ElementIntegralVariablePostprocessor'

### DISCUSSION THREAD

**Discussion by Leni-Yeo** [idaholab/moose#19977] [Q&A Modules: Phase field] [answered]
Good afternoon, I hope you are all doing well.

I am trying to do a bi-crystal simulation with an assigned energy and compare the value to that of the 'ElementIntegralVariablePostprocessor' output; using the 'AuxKernels' setup generally used in moose. I am using the 'PolycrystalKernel'  action for kernels setup. I am trying to have 'mu' as a constant input with a calculated grain boundary width. However, no matter the variation of 'mu', I am unable to have a match with both energies. I have tried using  'ElementIntegralVariablePostprocessor' for a number of setups where the grain boundary width is constant and 'mu' is calculated;  however, it always fails to yield the right value regardless of the inputs I have. 

I may have a wrong setup or missing a point. I was wondering if anyone has ever been able to have a match between output from 'ElementIntegralVariablePostprocessor' and the actual energy that is assigned in the input file ?. Please let me know if there is maybe a verified version/file in MOOSE that I may have missed too.

It would greatly help if you could also take a look at the files I have attached. I have an input file and its output file as well as the material class I quickly made to verify 'ElementIntegralVariablePostprocessor'.

Thank you in advance for your help.

[Files.zip](https://github.com/idaholab/moose/files/7830981/Files.zip)

**Comment by laagesen** [comment, marked as answer]
I'm not sure exactly what you mean by "assigning" an energy to the system. I see that you added a DerivativeParsedMaterial for the local free energy density that look likes it matches the Moelans model, and that you are adding it to the gradient term using the TotalFreeEnergy AuxKernel, passing it to an AuxVariable. Then you are integrating the AuxVariable using the ElementIntegralVariablePostprocessor. This all seems fine and should give you the correct integrated free energy for the entire system. However, I'm not sure what you are comparing that number to when you say you are "assigning" an energy, can you please elaborate? Are you trying to calculate the energy for this configuration analytically or something?

**Reply by Leni-Yeo** [reply]
Thank you for answering. 
Sorry for the misunderstanding. 
I am trying to compare the integrated free energy for the entire system using the ElementIntegralVariablePostprocessor to the sigma value that is inputted in the material block -- 0.708 J/m2.  It is my understanding that by multiplying the grain boundary area to the sigma value, we should be able to get the same total energy as what we get with the ElementIntegralVariablePostprocessor. However, both values are different for some reason. So, I was wandering if I may have a mistake in the setup or if it is an understanding issue. The main question is whether anyone was able to equate both results under any specific setup.

**Reply by laagesen** [reply]
The grain boundary energy value you gave is the energy per unit area of interface, so you are correct that you need to multiply by the area of the interface. It looks like you're using your own custom material property to calculate kappa. Can you do the same calculation using the `GBEvolution` material that comes with the framework instead to see if the results are consistent? The next thing to be aware of is that there is a length scale parameter in `GBEvolution` that sets the units for the problem. By default it is in nm, so if your simulation domain is set to be 300 by 300 in your input file, that means 300 nm by 300 nm. So I'm not sure if you did the same thing in your custom material property but that is another thing to check - you must use this length scale when you're determining the GB area for your analytical calculation of GB energy * GB area, which you are trying to match to the postprocessor calculation.

**Reply by Leni-Yeo** [reply]
Thank you.
I will try that and get back to you.

**Reply by Leni-Yeo** [reply]
Good afternoon, 
Sorry for the delay.

I ran a bi-crystal by modifying the built-in file "grain_growth_2D_voronoi.i". I attached the input / output files and a snip of the total energy from integrating the free energy for the entire system using the ElementIntegralVariablePostprocessor. 

- The total energy is at around "1337 eV". The mu value "1.89385" is the one calculated by "GBEvolution". I ran the simulation once at first and copied the value to the "DerivativeParsedMaterial" of the "free_energy" material. 

- The total energy when using sigma = 0.708 J/m^2 = 4.4 eV/nm^2 ; wGB = 14 nm; the height of the block  = 300 nm is: 4.4 * 14 * 300 =  18,480 eV. It is ~13.8 times greater.

I may have a mistake so I am not really 100% sure, but no matter the changes I tried both values are different somewhat.

[grain_growth_2D_voronoi_BICRYSTAL.zip](https://github.com/idaholab/moose/files/7907719/grain_growth_2D_voronoi_BICRYSTAL.zip)

**Reply by laagesen** [reply]
It looks like you are getting the correct result from the postprocessor but there is an error in your hand calculation. I think you may be getting confused by the fact that this is a 2D simulation, which is a "slice" of the real system in 3D. In the real 3D system, if the GB runs along the y-direction, then it sits in the yz plane and the x-direction is normal to the GB. So to calculate the energy of the GB, you would multiply sigma times the size in the y-direction times the size in the z-direction. You don't multiply by wGB, as this width (in the x-direction) is associated with the diffuse interface representation of the grain boundary. 

So assume the domain is 1 nm in the z-direction (the actual number doesn't matter). Your total energy for the analytical calculation is 4.4 eV/nm^2 * 300 nm * 1 nm = 1320 eV. To compare to the postprocessor, you need to recognize that the integrated value is also for a 2D domain which also has an arbitrary length in the z-direction; since we set it to 1 nm the total energy is 1337 eV/nm * 1 nm = 1337 eV. So you see the two vales are very close. You can repeat the exercise with any value you want for the length of the domain in the z-direction and you'll find the same thing. The factor of ~13.8 difference that you were getting is because you were multiplying the hand calculation by 14, which is incorrect.

You can probably make the postprocessor calculation match the hand calculation even more closely by increasing the resolution of the mesh in the interfacial region.

**Reply by Leni-Yeo** [reply]
Oh I see. It makes sense now.
I just did a trial run with my own material and the values are much closer.
Thank you so much. You were of great help.
